src/submit_jobs.py

```diff
@@ -123,17 +132,41 @@ def read_job(job: SparkJob):
         job.repartition()
 
     # Upload extracted files to AWS S3 in Hive-partitioned Parquet
+    aws = AWSClient(logger=session.logger)
+    new_local_files = []
     for job in jobs:
-        job.upload()
+        job_upload_results = job.upload(aws)
+        new_local_files.append(job_upload_results)
+
+    # If any jobs uploaded never-seen-before files, trigger a Glue crawler
+    if any(new_local_files):
```

Comment:
**[Suggestion, non-blocking]** I'm a little bit confused by the use of the `SparkJob.upload()` interface here. The fact that we're appending to `new_local_files` makes me think that `SparkJob.upload()` will be returning a string indicating a filename, but instead it returns a boolean. Not super high priority, but here are two alternatives that would be clearer to me:

1. Refactor `SparkJob.upload()` to return a list of strings representing the new local files that were uploaded, then adjust this block accordingly:

```suggestion
    new_local_files = []
    for job in jobs:
        new_local_files_from_job = job.upload(aws)
        if new_local_files_from_job:
            new_local_files += new_local_files_from_job

    # If any jobs uploaded never-seen-before files, trigger a Glue crawler
    if any(new_local_files):
```

2. Keep the `SparkJob.upload()` interface the same, but adjust this block to make it clear that `upload()` returns a bool:

```suggestion
    new_local_files = False
    for job in jobs:
        job_found_new_local_files = job.upload(aws)
        if job_found_new_local_files:
            new_local_files = True

    # If any jobs uploaded never-seen-before files, trigger a Glue crawler
    if new_local_files:
```

Both options work; I would prefer 1 if we think that it might be useful here or elsewhere to know the names of the new local files, and 2 otherwise.

Reply:
Good thinking! I went with 1 and also added a log line about the number of new files uploaded. See afdf098.